parsl/jobs/strategy.py

```diff
@@ -253,8 +249,9 @@ def _general_strategy(self, executor_facades, *, strategy_type):
                     if not self.executors[executor.label]['idle_since']:
                         logger.debug(f"Starting idle timer for executor. If idle time exceeds {self.max_idletime}s, blocks will be scaled in")
                         self.executors[executor.label]['idle_since'] = time.time()
-
                     idle_since = self.executors[executor.label]['idle_since']
+                    assert idle_since is not None, "The `if` statement above this assert should have forced idle time to be not-None"
+
```

Comment:
Would it work to force it with `: float = `?

Reply:
it's extracted from `self.executors[executor.label]['idle_since']` which is optional so you need to tell the type checker to disbelieve the type annotation: annotating idle_since as `float` would move the type error up to line 252, like this:

```
parsl/jobs/strategy.py:252: error: Incompatible types in assignment (expression has type "float | None", variable has type "float")  [assignment]
```

Reply:
you need to have some assert in here about "believe me... the types say you can have got an idle_since from this shared data structure that everyone can be poking at but I am promising you, it will be not None, even though I cannot prove some other thread did not modify the shared structure"

Reply:
(cannot prove *using the language of Python types...*)

Reply:
Yeah; can't keep the mypy idiosyncrasies clear sometimes.  In my mind, that's what `: float =` is doing.  "No, really, this is the case."  Okay.  It is what it is.

Reply:
:float is saying this is a place for a float so in two directions:

i) you can rely on the value to be a float (assuming you believe the annotation...)
ii) you should be sure you believe that what is put in that place is a float

The second of those is failing here: mypy has evidence that we can be assigning a None here, not a float (from the type signature of self.executors[executor.label]['idle_since'].

But because idle_since is a local variable (which self.executors[executor.label]['idle_since'] is not), we can believe that some other force isn't narrowing it down, and eliminate the None case.

If the `if` statement above had been working with locals rather than globals, I think it's possible mypy would have inferred that case-narrowing itself: there is more reasoning available about locals because they're not so prone to being poked at by code outside the current block.